main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala

```diff
@@ -457,6 +458,36 @@ object Simplifier {
               SimplifiedAst.Expr.Let(name, indexExp, exp, succ.tpe, exp.purity, loc)
           }
 
+        /**
+          * Matching a record may succeed or fail.
+          *
+          * We generate a fresh variable and let-binding for each component of the
+          * record (field or extension) and then we recurse on the nested patterns
+          * and freshly generated variables.
+          */
+        case (LoweredAst.Pattern.Record(pats, pat, tpe, loc) :: ps, v :: vs) =>
+          val freshVars = pats.map(_ => Symbol.freshVarSym("innerField" + Flix.Delimiter, BoundBy.Let, loc))
+          val fieldPats = pats.map(_.pat)
+          val varExp = SimplifiedAst.Expr.Var(v, tpe, loc)
+          val zero = patternMatchList(fieldPats ::: ps, freshVars ::: vs, guard, succ, fail)
+          // Note that we reverse pats and freshVars because we still want to fold right
+          // but we want to restrict the record / matchVar from left to right
+          val (one, restrictedMatchVar) = pats.reverse.zip(freshVars.reverse).foldRight((zero, varExp): (SimplifiedAst.Expr, SimplifiedAst.Expr)) {
+            case ((LoweredAst.Pattern.Record.RecordFieldPattern(field, _, pat, loc1), name), (exp, matchVarExp)) =>
+              val recordSelectExp = SimplifiedAst.Expr.ApplyAtomic(AtomicOp.RecordSelect(field), List(matchVarExp), pat.tpe, Pure, loc1)
+              val restrictedMatchVarExp = SimplifiedAst.Expr.ApplyAtomic(AtomicOp.RecordRestrict(field), List(matchVarExp), tpe, matchVarExp.purity, loc1)
```

Comment:
I suspect the type `tpe` (and also the purity) is wrong in `restrictedMatchVarExp`. It seems rather cumbersome to update the type with a record restriction based on how type rules works for `RecordRestrict`. However, this works for now. Thoughts?

Reply:
But conceptually you are right: it should be cumbersome. Because each RecordRestrict "peels of" one field.

But maybe a helper function would help?

Reply:
(and fields may not occur exactly in the order expected).

Reply:
> (and fields may not occur exactly in the order expected).

What do you mean? `{a, b, c}` does not always appear in the order `a :: b :: c :: Nil`?

Reply:
> I cannot comment on the details, but its very important that types and effects are correct for later compiler phases.
> 
> As for effects, it should be straightforward: There is only the effect of the scrutinee.
> 
> As for the types, you can try to enable the `Verifier` and see if it complains about your code specifically (it currently complains of other code).

How do I enable the `Verifier`? Is it a flag?

Reply:
I've created a helper function to make the record restriction type on the latest commit